channel2/classic_listener.go

```diff
@@ -78,70 +83,98 @@ func (listener *classicListener) Create() (Underlay, error) {
 
 	for {
 		select {
 		case peer := <-incoming:
-			impl := newClassicImpl(peer, 2)
-			if connectionId, err := globalRegistry.newConnectionId(); err == nil {
-				impl.connectionId = connectionId
-				request, hello, err := listener.receiveHello(impl)
-				if err == nil {
-					for _, h := range listener.handlers {
-						log.Infof("hello: %v, peer: %v, handler: %v", hello, peer, h)
-						if err := h.HandleConnection(hello, peer.PeerCertificates()); err != nil {
-							log.Errorf("connection handler error (%s)", err)
-							if err := listener.ackHello(impl, request, false, err.Error()); err != nil {
-								log.Errorf("error acknowledging hello (%s)", err)
-							}
-							break
-						}
-					}
-
-					impl.id = &identity.TokenId{Token: hello.IdToken}
-					impl.headers = hello.Headers
+			go listener.onConnect(peer)
+		case <-listener.close:
+			return
+		}
+	}
+}
 
-					if err := listener.ackHello(impl, request, true, ""); err == nil {
-						listener.created <- impl
-					} else {
-						log.Errorf("error acknowledging hello (%s)", err)
+func (listener *classicListener) onConnect(peer transport.Connection) {
+	logger := pfxlog.ContextLogger(listener.endpoint.String())
+	impl := newClassicImpl(peer, 2)
+	if connectionId, err := globalRegistry.newConnectionId(); err == nil {
+		impl.connectionId = connectionId
+		request, hello, err := listener.receiveHello(impl)
+		if err == nil {
+			for _, h := range listener.handlers {
+				logger.Infof("hello: %v, peer: %v, handler: %v", hello, peer, h)
+				if err := h.HandleConnection(hello, peer.PeerCertificates()); err != nil {
+					logger.Errorf("connection handler error (%s)", err)
+					if err := listener.ackHello(impl, request, false, err.Error()); err != nil {
+						logger.Errorf("error acknowledging hello (%s)", err)
 					}
-
-				} else {
-					log.Errorf("error receiving hello (%s)", err)
+					break
 				}
+			}
+
+			impl.id = &identity.TokenId{Token: hello.IdToken}
+			impl.headers = hello.Headers
+
+			if err := listener.ackHello(impl, request, true, ""); err == nil {
+				listener.created <- impl
 			} else {
-				log.Errorf("error getting connection id (%s)", err)
+				logger.Errorf("error acknowledging hello (%s)", err)
 			}
 
-		case <-listener.close:
-			return
+		} else {
+			logger.Errorf("error receiving hello (%s)", err)
 		}
+	} else {
+		logger.Errorf("error getting connection id (%s)", err)
 	}
+
+}
+
+type helloResponse struct {
+	message *Message
+	hello   *Hello
+	error   error
 }
 
 func (listener *classicListener) receiveHello(impl *classicImpl) (*Message, *Hello, error) {
-	log := pfxlog.ContextLogger(impl.Label())
-	log.Debug("started")
-	defer log.Debug("exited")
+	responseChan := make(chan helloResponse)
 
-	request, err := impl.rxHello()
-	if err != nil {
-		if err == UnknownVersionError {
-			writeUnknownVersionResponse(impl.peer.Writer())
+	go func() {
+		log := pfxlog.ContextLogger(impl.Label())
+		log.Debug("started")
+		defer log.Debug("exited")
```

Comment:
what happens when lots of these trigger all at the same time. the efficacy of these logs seem like they go down then. Maybe i'm missing something though but it might be worthwhile to just remove these now?

Reply:
Would lowering them to trace be better?

Reply:
is there a way to run with TRACE turned on?